### PR TITLE
[df] Cleanup execution artifacts from worker caches in Dask

### PR DESCRIPTION
**Pull request by vepadulano**
Every worker stores the RDataFrame created and the computation graph jitted during the first task in a process-wide dictionary serving as a cache. This is required in order to avoid having to re-jit the computation graph at each task.

This commit introduces a post-execution cleanup job for the Dask workers so that when a distributed execution ends, the workers are instructed to remove the artifacts generated by the execution from the caches.

In particular, if the RDataFrame execution was processing a TTree-based dataset, removing the artifact from the cache also notably deallocates the corresponding TTreeCache, one per RDataFrame in-flight in the process.

Here are some memory usage flamegraphs obtained with memray running the AGC analysis script

https://cernbox.cern.ch/s/YPEmBfULdvy8CbN

You will find the following:
* report_1sample_1file_2task_master.html : 1 RDataFrame, 1 file, 2 partitions. Total memory seen 293 MB
* report_1sample_1file_2task_cleancache.html : 1 RDataFrame, 1 file, 2 partitions. Total memory seen 238 MB
* report_5sample_1file_2task_master.html: 5 RDataFrames, 1 file each, 2 partitions each. Total memory seen 521 MB
* report_5sample_1file_2task_cleancache.html: 5 RDataFrames, 1 file each, 2 partitions each. Total memory seen 249 MB

One can search for the `TTreeCache` symbol and notice that e.g. between the second and the first flamegraphs the difference is ~40MB which is the size of the TTreeCache. Similarly between the 4th and 3rd there is approximately a difference of ~5 TTreeCache sizes.

### CODE REVIEW

**Comment by dpiparo** [comment]
Thanks a lot for this PR, that results from very intense debugging sessions initially enabled by @silverweed . Once the tests are green, I think these changes should be merged asap. There is no point in keeping around a fully jitted RDF+TTreeCache if it won't be used any more.

After the aforementioned step is completed, we might perhaps think about what can go wrong with used memory and what we can do about it. For example, suppose a workflow with N dataframes. For some reason we do not control (e.g. some Dask internals or something imposed by the infrastructure) the scheduling is such that all of them are alive until the end of the processing. Can we cap the memory used by the caches? For example we could just keep 1 buffer and assign it to the cache of the chain associated to the RDF pulled from the cache before using that RDF to process data (effectively using 1 single cache). De facto this means assuming N RDFs in flight but one TTreeCache buffer per worker. Apologies if I am missing something in the distrdf logic.

**Comment by vepadulano** [comment]
> Can we cap the memory used by the caches? For example we could just keep 1 buffer and assign it to the tree cache of the chain associated to the RDF pulled from the worker's rdfs cache before using that RDF to process data (effectively using 1 single cache). 

I believe this is at least in theory possible because we are in the situation that 1 worker == 1 core thus all the tasks the single worker processes are actually just sequential runs taken from a queue. The main obstacles on the way:

* This will need exposing quite a few TTree-only details from RDF, breaking the generic approach.
* This will need to be at the same time generic enough that it doesn't interfere with non-TTree-based executions (notably RNTuple).

**Comment by vepadulano** [comment]
> Once the tests are green, I think these changes should be merged asap.

Yes, I opened it to expose the fix and the result of the debugging session. While I polish it to avoid interfering with existing tests, this can be already tested on SWAN by injecting the custom DistRDF Python package taken from this branch on the PYTHONPATH of the workers, so we will try to get some measurements soon

**Comment by vepadulano** [comment]
The latest update of the commit removes the need for the extra swap of the internal RDF dataset with an empty one, improving on the first design. On my machine I can see the worker properly freeing up memory at the end of the execution of the analysis. It is still not ready for merging but we should now begin testing on SWAN+HTCondor to confirm the preliminary tests.

**Review by dpiparo — APPROVED** [review]
I understand from one of the comments that the changes are not ready. However, given their great effect, I would lean towards approving them. Improvements can be added at a later stage, after merging, if needed.

**Comment by vepadulano** [comment]
I have updated the contents of the CERNBox link from the PR description with more memray graphs, it looks clear that the effect of this PR is to properly remove the artifacts generated during the distributed execution, there is no more TTreeCache leaking leftover after the analysis (see any file ending with `_patch` and compare it with the same file ending with `_master`).

Now that the main source of memory usage is gone, I have noticed a much smaller contribute that still seems to make the memory increase in the flamegraphs. And it seems to have something to do with the increased number of files, at least on the surface. Let's take the following two files to compare:
* test_agc_9sample_1file_10task_patch.html
* test_agc_9sample_10file_10task_patch.html

They are running exactly the same script, with all 9 RDF samples. In the first case we have 1 file per sample, in the second case 10 files per sample. The memray flamegraph report a total of 253.5 MB used for the first case and 260.9 MB used for the first case (7.4 MB delta).

There are two main parts of the graph: the one deriving from the `clone` of the Dask process, where the actual analysis code is run, and the one deriving from the `_start` function where most of the Python code around the analysis is (things like `import`ing packages or the Dask worker-scheduler communication).

The difference between the two graphs in the `clone` part is of 0.6 MB, the difference between the two `_start` parts is 6.7 MB, so I focus on that from now on.

This 6.7 MB difference is in turn split between a 0.1 MB difference deriving from `_PyObject_VectorcallTState` and 6.6 MB difference deriving from. Finally, the vast majority of these 6.6 MB comes from:
* The metrics gathered by profiling (i.e. memray itself)
* Dask communication protocol

[Screencast from 2024-07-26 10-47-18.webm](https://github.com/user-attachments/assets/3a21ea3e-c130-42af-b50b-bb324b06eaaa)


TL;DR: After this patch the leftover memory increase may be attributed mostly to Dask or memray itself.

To confirm this, I have executed the full AGC benchmark on the SWAN platform with 64 workers of the CERN condor pools for 10 times in a row. At the end of the 10th run all the workers have a total RSS of < 1GB.

**Comment by dpiparo** [comment]
After this kind of analysis, I would propose we merge *now*.
